app/controllers/actions/RequiredUserAnswersForCorrectReturn.scala

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.actions
+
+import models.correctReturn.{AddASmallProducer, RepaymentMethod}
+import models.requests.DataRequest
+import models.{CheckMode, LitresInBands, NormalMode, RetrievedSubscription, SdilReturn}
+import pages.{Page, QuestionPage}
+import pages.correctReturn._
+import play.api.libs.json.Reads
+import play.api.mvc.Result
+import play.api.mvc.Results.Redirect
+import utilities.{GenericLogger, UserTypeCheck}
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+import scala.reflect.ClassTag
+
+class RequiredUserAnswersForCorrectReturn @Inject()(genericLogger: GenericLogger)(implicit val executionContext: ExecutionContext) extends ActionHelpers {
+
+  def requireData(page: Page)(action: => Future[Result])(implicit request: DataRequest[_]): Future[Result] = {
+    page match {
+      case CorrectReturnBaseCYAPage => checkYourAnswersRequiredData(action)
+      case CorrectReturnCheckChangesPage => checkChangesRequiredData(action)
+      case _ => action
+    }
+  }
+
+  private[controllers] def checkYourAnswersRequiredData(action: => Future[Result])(implicit request: DataRequest[_]): Future[Result] = {
+    val userAnswersMissing: List[CorrectReturnRequiredPage[_, _, _]] = returnMissingAnswers(mainRoute)
+    userAnswersMissing.headOption.map(_.pageRequired.asInstanceOf[Page]) match {
+      case Some(page) if List(PackAtBusinessAddressPage, PackagingSiteDetailsPage).contains(page) =>
+        genericLogger.logger.info(s"${request.userAnswers.id} now requires packaging sites")
+        Future.successful(Redirect(PackAtBusinessAddressPage.url(CheckMode)))
+      case Some(page) if page == SecondaryWarehouseDetailsPage =>
+        genericLogger.logger.info(s"${request.userAnswers.id} now has option to add warehouse")
+        Future.successful(Redirect(page.url(CheckMode)))
+      case Some(page) => genericLogger.logger.warn(s"${request.userAnswers.id} has hit correct return base CYA and is missing $userAnswersMissing")
+        Future.successful(Redirect(page.url(NormalMode)))
```

Comment:
Is it correct for it to be in NormalMode, rather than CheckMode? (I can think of reasons why it could be either)

Reply:
if the answers have not yet been completed then the user is in normal mode vs checking/changing answers in checked mode.